litex_boards/platforms/zcu104.py

```diff
@@ -0,0 +1,86 @@
+    ("user_btn", 1, Pins("C4"), IOStandard("LVCMOS33")),
+    ("user_btn", 2, Pins("B3"), IOStandard("LVCMOS33")),
+    ("user_btn", 3, Pins("C3"), IOStandard("LVCMOS33")),
+
+    ("serial", 0,
+        Subsignal("cts", Pins("A19")),
+        Subsignal("rts", Pins("C18")),
+        Subsignal("tx", Pins("C19")),
+        Subsignal("rx", Pins("A20")),
+        IOStandard("LVCMOS18")
+    ),
+
+    ("ddram", 0,
+        Subsignal("a", Pins(
+            "AH8 AG8 AF8 AG10 AG11 AH9 AG9 AH13",
+            "AK10 AJ10 AL8 AK8 AL12 AK12"),
+            IOStandard("SSTL12_DCI")),
+        Subsignal("ba", Pins("AL10 AL11"), IOStandard("SSTL12_DCI")),
+        Subsignal("bg", Pins("AE14 AE13"), IOStandard("SSTL12_DCI")),
+        Subsignal("ras_n", Pins("AF11"), IOStandard("SSTL12_DCI")),  # A16
+        Subsignal("cas_n", Pins("AE12"), IOStandard("SSTL12_DCI")),  # A15
+        Subsignal("we_n", Pins("AC12"), IOStandard("SSTL12_DCI")),   # A14
+        Subsignal("cs_n", Pins("AD12"), IOStandard("SSTL12_DCI")),
+        Subsignal("act_n", Pins("AD14"), IOStandard("SSTL12_DCI")),
+        Subsignal("alert_n", Pins("AF13"), IOStandard("SSTL12_DCI")),
+        Subsignal("par", Pins("AC13"), IOStandard("SSTL12_DCI")),
+        Subsignal("dm", Pins("AH18 AD15 AM16 AP18 AE18 AH22 AL20 AP19"),
+            IOStandard("POD12_DCI")),
+        Subsignal("dq", Pins(
+            "AG14 AG15 AF15 AF16 AF17 AE17 AG18 AF18",
+            "AD16 AD17 AB14 AA14 AB15 AB16 AC16 AC17",
+
+            "AJ15 AJ16 AK17 AJ17 AL18 AK18 AL15 AL16",
+            "AN16 AN17 AP15 AP16 AN18 AM18 AP13 AN13",
+
+            "AA20 AA19 AD19 AC18 AE20 AD20 AC19 AB19",
+            "AE24 AE23 AF22 AF21 AG20 AG19 AH21 AG21",
+
+            "AJ22 AJ21 AK20 AJ20 AK19 AJ19 AL23 AL22",
+            "AN23 AM23 AP23 AN22 AP22 AP21 AN19 AM19",
+            ),
+            IOStandard("POD12_DCI")),
+        Subsignal("dqs_p", Pins("AH14 AA16 AK15 AM14 AA18 AF23 AK22 AM21"),
+            IOStandard("DIFF_POD12_DCI")),
+        Subsignal("dqs_n", Pins("AJ14 AA15 AK14 AN14 AB18 AG23 AK23 AN21"),
+            IOStandard("DIFF_POD12_DCI")),
+        Subsignal("clk_p", Pins("AH11"), IOStandard("DIFF_SSTL12_DCI")),
+        Subsignal("clk_n", Pins("AJ11"), IOStandard("DIFF_SSTL12_DCI")),
+        Subsignal("cke", Pins("AB13"), IOStandard("SSTL12_DCI")),
+        Subsignal("odt", Pins("AF10"), IOStandard("SSTL12_DCI")),
+        Subsignal("reset_n", Pins("AF12"), IOStandard("LVCMOS12")),
+        Misc("SLEW=FAST"),
+    ),
+]
+
+# Platform -----------------------------------------------------------------------------------------
+
+class Platform(XilinxPlatform):
+    default_clk_name = "clk125"
+    default_clk_period = 1e9/125e6
+
+    def __init__(self):
+        XilinxPlatform.__init__(self, "xczu7ev-ffvc1156-2-i", _io, toolchain="vivado")
+
+    def create_programmer(self):
+        return VivadoProgrammer()
+
+    def do_finalize(self, fragment):
+        XilinxPlatform.do_finalize(self, fragment)
```

Comment:
From memory, I needed these lines here for reliable DDR4:
```
        self.add_platform_command("set_property INTERNAL_VREF {{0.84}} [get_iobanks 65]")
        self.add_platform_command("set_property INTERNAL_VREF {{0.84}} [get_iobanks 66]")
```

Reply:
indeed, it will not work without it.

Reply:
indeed, it will not work without it.